dali/pipeline/executor/executor_impl.cc

```diff
@@ -748,8 +751,15 @@ Checkpoint &Executor<WorkspacePolicy, QueuePolicy>::GetCurrentCheckpoint() {
 template<typename WorkspacePolicy, typename QueuePolicy>
 void Executor<WorkspacePolicy, QueuePolicy>::RestoreStateFromCheckpoint(const Checkpoint &cpt) {
   DALI_ENFORCE(cpu_iteration_id_ == 0, "Cannot restore state of a running executor. ");
-  for (int i = 0; i < graph_->NumOp(); ++i)
-    graph_->Node(i).op->RestoreState(cpt.GetOpCheckpoint(i));
+
+  for (int i = 0, n = cpt.NumOp(); i < n; i++) {
+    const auto &op_cpt = cpt.GetOpCheckpoint(i);
+    auto *op = GetOperator(op_cpt.OperatorName());
+    DALI_ENFORCE(op != nullptr, make_string(
+        "Operator with name \"", op_cpt.OperatorName(), "\" "
+        "from the checkpoint not found in the graph."));
+    op->RestoreState(op_cpt);
+  }
```

Comment:
There's a slight change in what checkpoints we accept now, namely, if the pipeline has some extra ops that are not present in the checkpoint, we will silently accept it and start with some of the ops, potentially, not restored.

How about iterating here over the graph and getting the OpCheckpoints by name? Together with the deserialization part we would check inclusions in both direction.

The separate issue is whether we may want behaviour of this kind, for example, explictly allowing for no OpCheckpoint for stateless operators. But that deserves a follow-up, what I am worried about here is silently loosening the validation.

Reply:
Added a check that the number of operators matches.